Feature/events

dillag opened the pull request:
ready to review and to merge< add tree for auth user, add trees for other users, add crud events

dillag committed:
CRUD events

dillag committed:
new models and views

dillag committed:
ready to merge, add tree for auth user, add trees for other users, add crud events

AVVasiliev commented on BootCamp/roadmap/serializers.py:
```diff
@@ -0,0 +1,34 @@
+
+    def create(self, validated_data):
+        validated_data["user_id"] = User.objects.filter(username=validated_data["user_id"])[0]
+        validated_data["node_id"] = Node.objects.filter(name=validated_data["node_id"])[0]
```
какой-то разброс в именах параметров, node_id это name? или это id чего-то? Аналогично с username и user_id

AVVasiliev commented on BootCamp/roadmap/serializers.py:
```diff
@@ -0,0 +1,34 @@
+    node_id = serializers.CharField(max_length=150, allow_blank=True)
+    status = serializers.CharField(max_length=300, allow_blank=True)
+
+    def create(self, validated_data):
```
Этим не надо пользоваться, были даны примеры, как это делать самостоятельно, нужно убрать. Вся работа по созданию/изменению с моделями идет во view, сериализатор не связан с моделью - мы так договорились

dillag replied:
исправил

AVVasiliev commented on BootCamp/roadmap/serializers.py:
```diff
@@ -0,0 +1,34 @@
+
+    def update(self, instance, validated_data):
+        print(type(instance))
+        print(validated_data)
```
Удалить

dillag replied:
исправил

AVVasiliev commented on BootCamp/roadmap/urls.py:
```diff
@@ -0,0 +1,13 @@
+urlpatterns = [
+    path('', views.index, name='index'),
+    path('<int:pk>', views.indexUsers, name='indexUser'),
+    path('api/v1/test/myevents', ViewEvent.as_view()),
```
что за test? к api/v1 можно через include добавляить урлы, а не писать с 0

dillag replied:
тоже исправил

AVVasiliev commented on BootCamp/roadmap/urls.py:
```diff
@@ -0,0 +1,13 @@
+from django.urls import path
+
+from . import views
+from .views import *
```
лучше использовать явные импорты: from views import View1, View2

dillag replied:
и это тоже

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,126 @@
+            serializer = EventListSerializer(query, many=True)
+            return Response(serializer.data)
+        else:
+            return Response({'error': 'error'})
```
очень ясно и понятно что за ошибка

dillag replied:
вроде более понятно расписал

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,126 @@
+    serializer_class = EventSerializer
+
+    def get(self, request):
+        if request.user.is_authenticated:
```
ваша ветка разве связана с авторизацией? это здесь не нужно, это делается не ручной проверкой в каждом запросе

dillag replied:
убрал

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,126 @@
+            return Response({'error': 'error'})
+
+    def put(self, request, pk):
+        print(request.user)
```
лишнее

dillag replied:
тоже убрал

AVVasiliev commented on BootCamp/roadmap/migrations/0006_alter_event_enddate.py:
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 4.0.2 on 2022-02-21 13:36
```
К чему так много миграций? Этих изменений ни у кого нет, это все можно смело откатить, и сделать одну нормальную миграцию

dillag replied:
исправил

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,126 @@
+
+    def post(self, request):
+        if request.user.is_authenticated:
+            user = User.objects.filter(username=request.user.username)[0]
```
Почему не User.objects.get()??

dillag replied:
заменил

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,126 @@
+    def get(self, request):
+        if request.user.is_authenticated:
+            user = User.objects.filter(username=request.user.username)[0]
+            query = Event.objects.filter()
```
Что фильтруем?

dillag replied:
уже ничего

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,126 @@
+    def put(self, request, pk):
+        if request.user.is_authenticated:
+            user = User.objects.filter(username=request.user.username)[0]
+            query = get_object_or_404(Event.objects.all(), pk=pk)
```
штука интересная, но не явная - переделать на явную, и опять таки вещи не особо корректная - 404 это значит нет запроса по такому адресу, а запрос же есть, он обработал данные и дал ответ, тут уместно будет 400 или 409

dillag replied:
переделал

dillag committed:
ready to merge, add tree for auth user, add trees for other users, add crud events
new:
delete CrUd from serializer and add them to views,
clear migrations,
add "include" to urls,
clear import,
delete all prints,
changed "filter" to "get",
add "try...except",
refactor

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,140 @@
+    def get(self, request):
+        try:
+            user = User.objects.get(username=request.user.username)
+        except Exception:
```
Нужно всегда стараться отлавливать конкретные исключения - здесь можно предположить  что в качестве исключений может быть User.DoesNotExists и User.MultipleObjectReturned
Еще интересно почему мы получаем пользователя по username - оно не выглядит прям уникальным полем

dillag replied:
исправил

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,140 @@
+        good_list = [i.node_id.name for i in good_query]
+
+        ready_query = Event.objects.filter(user_id=user, status=1)
+        ready_list = [i.node_id.name for i in ready_query]
```
Это все можно сделать сразу через orm иcпользуя values_list и annotate

dillag replied:
переложил всю ответственность на орм

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,140 @@
+            user = User.objects.get(pk=pk)
+        except Exception:
+            return Response({"error": "Пользователь не найден"}, status=409)
+        good_query = Event.objects.filter(user_id=user, status=2)
```
для статусов лучше заводить Enum класс в котором будут с числами явно ассоциированы значения, например:
```
from django.db.models import IntegerChoices

class TestChoices(IntegerChoices):
    success = "success", 0
    failed = "failed", 1
```

dillag replied:
завел класс, а он меня нет(

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,140 @@
+    def put(self, request, pk):
+        try:
+            query = Event.objects.get(pk=pk)
+        except Exception:
```
Опять таки мы можем предвидеть какие исключение падают - а значит можем отловить конкретные

dillag replied:
теперь конкретнее некуда

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,140 @@
+            return Response({"error": "Пользователь не найден"}, status=400)
+        events = request.data
+        serializer = EventSerializer(data=events)
+        if serializer.is_valid(raise_exception=True):
```
if здесь не нужен, код дальше не пойдет так как стоит raise_exception=True

dillag replied:
убрал

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,140 @@
+    def delete(self, request, pk):
+        event = Event.objects.get(pk=pk)
+        event.delete()
+        return Response({"message": "Удалил".format(pk)}, status=204)
```
Синтаксическая ошибка

dillag replied:
починил

AVVasiliev commented on BootCamp/roadmap/views.py:
```diff
@@ -0,0 +1,140 @@
+            return Response(serializer.data)
+
+    def delete(self, request, pk):
+        event = Event.objects.get(pk=pk)
```
А если мы передали неправильный pk? Нужно обработать этот сценарий

dillag replied:
обработал

dillag committed:
ready to merge, add tree for auth user, add trees for other users, add crud events
new:
delete CrUd from serializer and add them to views,
clear migrations,
add "include" to urls,
clear import,
delete all prints,
changed "filter" to "get",
add "try...except",
refactor
add choice.py,
add values_list,
add new except

dillag committed:
ready to merge, add tree for auth user, add trees for other users, add crud events
new:
delete CrUd from serializer and add them to views,
clear migrations,
add "include" to urls,
clear import,
delete all prints,
changed "filter" to "get",
add "try...except",
refactor
add choice.py,
add values_list,
add new except

dillag committed:
ready to merge, add tree for auth user, add trees for other users, add crud events
new:
delete CrUd from serializer and add them to views,
clear migrations,
add "include" to urls,
clear import,
delete all prints,
changed "filter" to "get",
add "try...except",
refactor
add choice.py,
add values_list,
add new except,
use choice.py:)

dillag committed:
Merge remote-tracking branch 'origin/main' into feature/events

# Conflicts:
#	BootCamp/BootCamp/settings.py
#	BootCamp/roadmap/admin.py
#	BootCamp/roadmap/apps.py
#	BootCamp/roadmap/models.py
#	BootCamp/roadmap/tests.py
#	BootCamp/roadmap/views.py

lydiym commented:
Конфликты не позволяют провести ревью.

dillag committed:
ready to merge, add tree for auth user, add trees for other users, add crud events
new:
delete CrUd from serializer and add them to views,
clear migrations,
add "include" to urls,
clear import,
delete all prints,
changed "filter" to "get",
add "try...except",
refactor
add choice.py,
add values_list,
add new except,
use choice.py:)
